File: ios-security/EmulatorDetection.m
#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import <sys/sysctl.h>
#import <mach-o/dyld.h>

#pragma mark - Native checks

static BOOL isSimulator(void) {
#if TARGET_OS_SIMULATOR
    return YES;
#else
    return NO;
#endif
}

static BOOL isDebugged(void) {
    int name[4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid() };
    struct kinfo_proc info;
    size_t size = sizeof(info);
    memset(&info, 0, sizeof(info));
    sysctl(name, 4, &info, &size, NULL, 0);
    return ((info.kp_proc.p_flag & P_TRACED) != 0);
}

static BOOL isJailbroken(void) {
#if TARGET_OS_SIMULATOR
    return YES;
#endif

    NSArray *paths = @[
        @"/Applications/Cydia.app",
        @"/Library/MobileSubstrate/MobileSubstrate.dylib",
        @"/bin/bash",
        @"/usr/sbin/sshd",
        @"/etc/apt",
        @"/private/var/lib/apt/"
    ];

    NSFileManager *fm = [NSFileManager defaultManager];
    for (NSString *p in paths) {
        if ([fm fileExistsAtPath:p]) return YES;
    }

    NSError *err = nil;
    NSString *testPath = @"/private/jb_test.txt";
    [@"test" writeToFile:testPath atomically:YES encoding:NSUTF8StringEncoding error:&err];
    if (err == nil) {
        [fm removeItemAtPath:testPath error:nil];
        return YES;
    }

    return NO;
}

static BOOL isFridaDetected(void) {
    NSArray *keywords = @[
        @"frida",
        @"frida-gadget",
        @"libfrida",
        @"gum-js-loop",
        @"re.frida",
        @"linjector"
    ];

    uint32_t count = _dyld_image_count();
    for (uint32_t i = 0; i < count; i++) {
        const char *cname = _dyld_get_image_name(i);
        if (!cname) continue;

        NSString *name = [[NSString alloc] initWithUTF8String:cname];
        NSString *lower = name.lowercaseString;

        for (NSString *k in keywords) {
            if ([lower containsString:k]) {
                return YES;
            }
        }
    }
    return NO;
}

#pragma mark - AUTO EXECUTION (NO AppDelegate)

__attribute__((constructor))
static void native_security_gate(void) {
#if DEBUG
    return;
#endif

    if (isSimulator() ||
        isDebugged() ||
        isJailbroken() ||
        isFridaDetected()) {

        dispatch_async(dispatch_get_main_queue(), ^{
            exit(0);
        });
    }
}

File: ios-security/EmulatorDetection.swift
import Foundation
import UIKit
import React

@objc(EmulatorDetection)
class EmulatorDetection: NSObject {

  @objc
  func isEmulator(_ resolve: RCTPromiseResolveBlock,
                  rejecter reject: RCTPromiseRejectBlock) {

    #if TARGET_IPHONE_SIMULATOR
      resolve(true)   // Simulator = Emulator
      return
    #endif

    resolve(Self.isJailbroken())
  }

  @objc
  func isDebugging(_ resolve: RCTPromiseResolveBlock,
                   rejecter reject: RCTPromiseRejectBlock) {

      var name = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
      var info = kinfo_proc()
      var size = MemoryLayout<kinfo_proc>.stride

      let sysctlResult = sysctl(&name, UInt32(name.count), &info, &size, nil, 0)

      let isDebugged = (sysctlResult == 0) && ((info.kp_proc.p_flag & P_TRACED) != 0)

      resolve(isDebugged)
  }
  
  static func isJailbroken() -> Bool {
    return hasSuspiciousFiles() ||
           canOpenCydia() ||
           canWriteOutsideSandbox()
  }

  private static func hasSuspiciousFiles() -> Bool {
    let paths = [
      "/Applications/Cydia.app",
      "/Library/MobileSubstrate/MobileSubstrate.dylib",
      "/bin/bash",
      "/usr/sbin/sshd",
      "/etc/apt"
    ]
    return paths.contains { FileManager.default.fileExists(atPath: $0) }
  }

  private static func canOpenCydia() -> Bool {
    guard let url = URL(string: "cydia://package/com.example") else { return false }
    return UIApplication.shared.canOpenURL(url)
  }

  private static func canWriteOutsideSandbox() -> Bool {
    let test = "/private/jb_test.txt"
    do {
      try "test".write(toFile: test, atomically: true, encoding: .utf8)
      try? FileManager.default.removeItem(atPath: test)
      return true
    } catch {
      return false
    }
  }
}
